src/lib.rs

```diff
@@ -1641,6 +1691,219 @@ impl Instance {
+                                );
+                                let from_handle = from_table.remove(*handle)?;
+                                let to_handle = table_array[to_rid].add(HandleElement {
+                                    rep: from_handle.rep,
+                                    own: true,
+                                    lend_count: 0,
+                                });
+                                *result = to_handle;
+                                Ok(())
+                            },
+                        )))
+                    }
+                    GeneratedTrampoline::ResourceTransferBorrow => {
+                        let ty = ty.with_name("resource-transfer-borrow");
+                        let tables = inner.state_table.clone();
+                        Ok(Extern::Func(wasm_runtime_layer::Func::new(
+                            ctx.as_context_mut().inner,
+                            ty,
+                            move |_ctx, args, results| {
+                                let (handle, from_rid, to_rid) = match args {
+                                    [
+                                        wasm_runtime_layer::Value::I32(handle),
+                                        wasm_runtime_layer::Value::I32(from_rid),
+                                        wasm_runtime_layer::Value::I32(to_rid),
+                                    ] => (handle, from_rid, to_rid),
+                                    args => bail!(
+                                        "resource-transfer-borrow(handle: i32, from-rid: i32, to-rid: i32)\
+                                         called with unexpected args {:?}", args
+                                    ),
+                                };
+                                let from_rid = usize::try_from(*from_rid)?;
+                                let to_rid = usize::try_from(*to_rid)?;
+                                let result = match results {
+                                    [wasm_runtime_layer::Value::I32(result)] => result,
+                                    results => bail!(
+                                        "resource-transfer-borrow(handle: i32, from-rid: i32, to-rid: i32)\
+                                         call expects unexpected results {:?}", results
+                                    ),
+                                };
+                                let mut table_array = tables
+                                    .resource_tables
+                                    .try_lock()
+                                    .expect("Could not get mutual reference to table.");
+                                let from_table = &mut table_array[from_rid];
+                                let handle_borrow = from_table.get(*handle)?;
+                                let handle_rep = handle_borrow.rep;
+                                // FIXME: wrong condition, should be if from_rid is imported
```

Comment:
@DouglasDwyer I am unsure about how to implement this part (and about whether I understand Jco's code here https://github.com/bytecodealliance/jco/blob/1420aad1679418da6037bb787a9130b88453bcd3/crates/js-component-bindgen/src/intrinsics.rs#L198-L211). Jco checks whether the resource is imported or local to this instance - is this information already stored somewhere / how could I get it?

Reply:
Hey @juntyr, thanks again for your hard work on this! I do not believe that I store whether the resource is imported or local to an instance anywhere currently.

I would need to do a bit more reading to understand why the handle is not cleared if the resource is local to the instance. Why would the behavior be different?

I am wondering if this check has to do with the "simplification" where resources within the same instance aren't tracked? That simplification isn't implemented at all in this crate right now.

Reply:
I hadn’t considered that “simplification” before but I think you’re right that this might be the reason for this if statement in Jco. Without it, would a borrow transfer just decrease the lend count in the “from” resource and create a new resource instance for “to”? Is this the purpose of the lend count? If we always create a new resource instance, we also loose the second if statement which just uses the inner representation inside the defining instance, otherwise we might gain layers of indirection which could mess with that instance‘s assumption that the representation is just a pointer?